app/controllers/application_controller.rb

```diff
@@ -104,4 +105,20 @@ def is_valid_signed_url?(signed_url)
     # return true if the scheme is https, the hostname matches a known GCS host, and the query string parameters have the correct keys
     uri.scheme == 'https' && ValidationTools::GCS_HOSTNAMES.include?(uri.hostname) && parsed_query.keys == ValidationTools::SIGNED_URL_KEYS
   end
+
+  def handle_401
+    respond_to do |format|
+      format.html {redirect_to new_user_session_path, alert: 'Your session has expired.'}
+      format.js {render js: "alert('Your session has expired; Please log in again')"}
+    end
+  end
+
+  protected
+
+  # patch to supply CSRF token on all ajax requests if it is not present
+  def set_csrf_headers
+    if request.xhr?
+      cookies['XSRF-TOKEN'] = form_authenticity_token if cookies['XSRF-TOKEN'].blank?
```

Comment:
Could you explain what service is expecting `XSRF-TOKEN`?  Knowing that would help me remember which services call this a "CSRF" and which call it a "XSRF".

Reply:
That's a great question - what I've been able to deduce about this is that it's a Rails-specific header for CSRF tokens on AJAX requests.